paywall/src/windowTypes.ts

```diff
@@ -0,0 +1,123 @@
+import { PaywallConfig } from './unlockTypes'
+
+// window sub-types
+
+export interface EventDocument {
+  createEvent: (type: string) => CustomEvent
+}
+
+export interface IframeManagingDocument {
+  createElement: (type: 'iframe') => IframeType
+  querySelector: (selector: string) => any
+  body: {
+    insertAdjacentElement: (where: 'afterbegin', iframe: IframeType) => void
+    style: {
+      overflow: string
+    }
+  }
+}
+
+export interface FullDocument extends EventDocument, IframeManagingDocument {}
+
+// used in unlock.js/dispatchEvent.ts
+export interface EventWindow {
+  CustomEvent: {
+    // this is copy/paste from the .d.ts for window, because it
+    // declares CustomEvent as an interface and exports it, but
+    // also as a global variable, and the global variable is not
+    // also exported in the definition of window.
+    // So, calling new window.CustomEvent() fails, but
+    // new CustomEvent() succeeds.
+    prototype: CustomEvent
+    new <T>(typeArg: string, eventInitDict?: CustomEventInit<T>): CustomEvent<T>
+  }
+  document: EventDocument
+  dispatchEvent: (event: CustomEvent) => void
+}
+
+// used anywhere cache is used
+export interface LocalStorageWindow {
+  localStorage: Storage
+}
+
+// used in web3Proxy.ts
+export interface web3MethodCall {
+  method: string
+  params: any[]
+  jsonrpc: '2.0'
+  id: number
+}
+
+export type web3Callback = (error: Error | string | null, result: any) => void
+export type web3Send = (
+  methodCall: web3MethodCall,
+  callback: web3Callback
+) => void
+
+export interface Web3Window extends PostOfficeWindow {
+  web3?: {
+    currentProvider: {
+      sendAsync?: web3Send
+      send?: web3Send
+      isMetamask: true | undefined
+    }
+  }
+}
+
+// used in utils/postOffice.ts
+export interface MessageEvent {
+  source: any
+  origin: string
+  data: any
+}
+
+export type MessageHandler = (event: MessageEvent) => void
+
+export interface PostOfficeWindow {
+  addEventListener: (type: 'message', handler: MessageHandler) => void
+}
+
+export interface IframePostOfficeWindow extends PostOfficeWindow {
+  parent: PostMessageTarget
+  location: {
+    href: string
+  }
+}
+
+export interface PostMessageTarget {
+  postMessage: (data: any, origin: string) => void
+}
+
+export interface IframeType {
+  contentWindow: PostMessageTarget
+  className: string
+  src: string
+  setAttribute: (attr: 'src', value: string) => void
```

Comment:
In scenarios like this, should this actually be a method with a single parameter `value`?

Reply:
I'm confused, this is the actual signature of setAttribute, we can't change what `window` provides here

Reply:
what I am doing is making sure we only call `setAttribute` with `'src'` since that is the only attribute we care about in `unlock.1.0.min.js`

Reply:
ah okay!

Reply:
talked to Chris about this, he had an idea to make it clearer I'm going to implement:

```ts
type AttributeName = 'src' /* can extend with a union of | 'anotherattribute' if needed */
interface Blah {
  setAttribute: (attr: AttributeName, value: string) => void
}
```